staging/src/k8s.io/component-base/logs/spancontextlogger/spancontextlogger.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package spancontextlogger includes utility functions for correlating logs
+// with traces.
+package spancontextlogger
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/trace"
+	"k8s.io/klog/v2"
+)
+
+// LoggerWithSpanContext adds trace_id and span_id values to the logger if the
+// context includes a span context.
+func LoggerWithSpanContext(ctx context.Context, logger klog.Logger) klog.Logger {
```

Comment:
Where would this function get called?

Reply:
The intent was that that this would be called at each log statement we care about e.g. (randomly chosen) https://github.com/kubernetes/kubernetes/blob/40f222b6201d5c6476e5b20f57a4a7d8b2d71845/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L408

Something like:

```golang
spancontextlogger.LoggerWithSpanContext(ctx, klog.Background()).V(4).Infof(...)
```

If we had https://github.com/kubernetes/klog/pull/415, we could either add this to the trace.StartSpan call (we would have to wrap the OTel SDK), or we could add it in `klog.FromContext` (which would require klog to take a dependency on go.opentelemetry.io/otel/trace).

Then the typical contextual logging usage would work:

```golang
klog.FromContext(ctx).V(4).Infof(...)
```

Reply:
> The intent was that that this would be called at each log statement we care about e.g. (randomly chosen) 

So either we convert many existing log calls (potentially adding the otel dependency to a package if it didn't before!) or we convert only a few calls, but then correlation between a trace and the corresponding log output is incomplete. Neither seems very appealing.

> we could either add this to the trace.StartSpan call (we would have to wrap the OTel SDK)

In other words, whenever a span and/or trace ID gets added to a context, also add a modified logger.

> we could add it in klog.FromContext (which would require klog to take a dependency on go.opentelemetry.io/otel/trace).

Not necessarily. We could implement a hook mechanism in klog such that by default, `FromContext` works as before. In Kubernetes components we can then hook into `FromContext` and add this special otel support. Typically the package which sets up logging will already depend on otel, so nothing changes.

It's a bit hard to predict which approach will yield better performance. It depends on how often code calls `FromContext` and how often IDs get added. I suspect we need to try out both.

Which component should we target with this? Does it have good performance benchmarks?

We could use scheduler_perf, but I suspect that it doesn't really do much with otel.

Reply:
> It depends on how often code calls FromContext and how often IDs get added. I suspect we need to try out both.

For log levels (e.g. debug) that won't be enabled in production, will FromContext still be called even if the line will not be logged?  If not, I would expect StartSpan to be called much more often, given that it is called even when the span is not sampled.  If it is, then i'm less sure.  It would depend on how much debug logging a component has, which probably differs a lot.

Reply:
It also depends on how expensive deduplication is compared with adding trace_id and span_id.

Reply:
> will FromContext still be called even if the line will not be logged

Yes, because the code has to have the logger before it can ask the logger whether something gets logged.

>  If not, I would expect StartSpan to be called much more often, given that it is called even when the span is not sampled.

If the span is not sampled, we wouldn't modify the logger, would we?

Reply:
At least in OTel logging, the trace ID, span ID, and trace flags are added to all log lines, not just sampled ones.  It can be useful to be able to search for a trace ID in logs, even for non-sampled spans.